ping/ping.c

```diff
@@ -665,39 +705,13 @@ int ping4_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 
 		if (probe_fd < 0)
 			error(2, errno, "socket");
+
 		if (rts->device) {
-			struct ifreq ifr;
 			int i;
 			int fds[2] = {probe_fd, sock->fd};
 
-			memset(&ifr, 0, sizeof(ifr));
-			strncpy(ifr.ifr_name, rts->device, IFNAMSIZ - 1);
-
-			for (i = 0; i < 2; i++) {
-				int fd = fds[i];
-				int rc;
-				int errno_save;
-
-				enable_capability_raw();
-				rc = setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE,
-						rts->device, strlen(rts->device) + 1);
-				errno_save = errno;
-				disable_capability_raw();
-
-				if (rc == -1) {
-					if (IN_MULTICAST(ntohl(dst.sin_addr.s_addr))) {
-						struct ip_mreqn imr;
-						if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0)
-							error(2, 0, _("unknown iface: %s"), rts->device);
-						memset(&imr, 0, sizeof(imr));
-						imr.imr_ifindex = ifr.ifr_ifindex;
-						if (setsockopt(fd, SOL_IP, IP_MULTICAST_IF,
-							       &imr, sizeof(imr)) == -1)
-							error(2, errno, "IP_MULTICAST_IF");
-					} else
-						error(2, errno_save, "SO_BINDTODEVICE %s", rts->device);
-				}
-			}
+			for (i = 0; i < 2; i++)
+				bind_to_device(rts, fds[i], dst.sin_addr.s_addr);
```

Comment:
I guess that we can get rid of the fds[] array as well, since we have put the code into the function we can do:
```
bind_to_device(rts, probe_fd, dst.sin_addr.s_addr);
bind_to_device(rts, sock->fd, dst.sin_addr.s_addr);
```

Other than that the code looks good to me.